pkg/syncer/fullsync.go

```diff
@@ -320,6 +365,263 @@ func CsiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer, vc s
+	var terms []v1.NodeSelectorTerm
+	for _, topology := range accessibleTopology {
+		if len(topology.Segments) == 0 {
+			continue
+		}
+
+		var expressions []v1.NodeSelectorRequirement
+		for k, v := range topology.Segments {
+			expressions = append(expressions, v1.NodeSelectorRequirement{
+				Key:      k,
+				Operator: v1.NodeSelectorOpIn,
+				Values:   []string{v},
+			})
+		}
+		terms = append(terms, v1.NodeSelectorTerm{
+			MatchExpressions: expressions,
+		})
+	}
+
+	return &v1.VolumeNodeAffinity{
+		Required: &v1.NodeSelector{
+			NodeSelectorTerms: terms,
+		},
+	}
+}
+
+// generateVolumeAccessibleTopologyJSON returns JSON string from AccessibleTopology given as input.
+// This value will be set on the PVC for annotation - "csi.vsphere.volume-accessible-topology"
+func generateVolumeAccessibleTopologyJSON(topologies []*csi.Topology) (string, error) {
+	segmentsArray := make([]string, 0)
+	for _, topology := range topologies {
+		jsonSegment, err := json.Marshal(topology.Segments)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal topology segment: %v to json. Err: %v", topology.Segments, err)
+		}
+		segmentsArray = append(segmentsArray, string(jsonSegment))
+	}
+	return "[" + strings.Join(segmentsArray, ",") + "]", nil
+}
+
+// patchNodeAffinityToPVAndPVC finds the topology associated with PV and patches that info as
+// node affinity to PV objects
+// This also adds "csi.vsphere.volume-accessible-topology" annotation to associated PVC, if any
+func patchNodeAffinityToPVAndPVC(ctx context.Context, metadataSyncer *metadataSyncInformer,
+	vc *cnsvsphere.VirtualCenter, pvWithoutNodeAffinity []*v1.PersistentVolume,
+	pvToPVCMap map[string]*v1.PersistentVolumeClaim) error {
+	log := logger.GetLogger(ctx)
+	// Iterate over all PVs missing node affinity info, discover the topology and patch to both PV & PVC
+	for _, pv := range pvWithoutNodeAffinity {
+		log.Infof("Setting node affinity for pv: %q", pv.Name)
+		oldData, err := json.Marshal(pv)
+		if err != nil {
+			log.Errorf("failed to marshal pv: %v, Error: %v", pv, err)
+			continue
+		}
+		pvCSITopology, err := getPVNodeAffinity(ctx, metadataSyncer, vc, pv)
+		if err != nil {
+			log.Errorf("unable to get node affinity for PV %q. Error: %+v",
+				pv.Name, err)
+			return err
+		}
+		newPV := pv.DeepCopy()
+		newPV.Spec.NodeAffinity = GenerateVolumeNodeAffinity(pvCSITopology)
+		newData, err := json.Marshal(newPV)
+		if err != nil {
+			log.Errorf("failed to marshal updated PV with node affinity rules: %v, Error: %v", newPV, err)
+			continue
+		}
+		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, pv)
+		if err != nil {
+			log.Errorf("error Creating two way merge patch for PV %q with error : %v", pv.Name, err)
+			return err
+		}
+		k8sClient, err := k8s.NewClient(ctx)
+		if err != nil {
+			log.Errorf("Failed to create kubernetes client. Err: %+v", err)
+			return err
+		}
+		attempt := 0
+		// Patch node affinity to PV
+		for {
+			attempt++
```

Comment:
remove re-try attempts to patch PVs and PVCs

Reply:
done